src/Mapper/MMC1.re

```diff
@@ -0,0 +1,88 @@
+type prg_mode =
+  | SwitchBoth
+  | SwitchHigh
+  | SwitchLow;
+type chr_mode =
+  | Switch1
+  | Switch2;
+
+let mmc1 = (rom: Rom.t): Mapper.t => {
+  let mirroring = ref(rom.mirroring);
+  let prg_bank = ref(0);
+  let chr_bank0 = ref(0);
+  let chr_bank1 = ref(0);
+  let write_count = ref(0);
+  let accumulator = ref(0);
+  let prg_mode = ref(SwitchBoth);
+  let chr_mode = ref(Switch1);
+
+  let chr_addr = offset => {
+    let bank = offset < 0x1000 ? chr_bank0^ : chr_bank1^;
+    bank * 0x1000 + offset land 0xfff;
+  };
+  let update = address =>
+    if (address < 0xa000) {
+      mirroring :=
+        (
+          switch (address land 0x3) {
+          | 0 => Lower
+          | 1 => Upper
+          | 2 => Vertical
+          | _ => Horizontal
+          }
+        );
+      prg_mode :=
+        (
+          switch (address lsr 2 land 0x3) {
+          | 2 => SwitchHigh
+          | 3 => SwitchLow
+          | _ => SwitchBoth
+          }
+        );
+      chr_mode :=
+        (
+          switch (address lsr 4 land 0x1) {
+          | 0 => Switch2
+          | _ => Switch1
+          }
+        );
+    } else if (address < 0xc000) {
+      if (chr_mode^ == Switch1) {
+        chr_bank0 := accumulator^;
+      } else {
+        chr_bank0 := accumulator^;
+        chr_bank1 := accumulator^ land 1;
```

Comment:
See the comment in tests; I'm not entirely sure how this is supposed to work.

Reply:
Ohh, I should re-read the docs again. Maybe only `chr_bank1` is wrapped and `chr_bank0` isn't? My approach last time was "just keep the programmers from trying to pick a CHR bank that doesn't exist which (at least for a first release) seems fine to me.